test/e2e/dynamic_provisioning_test.go

```diff
@@ -1652,7 +1725,7 @@ func (t *dynamicProvisioningTestSuite) defineTests(isMultiZone bool, schedulerNa
 		storageClassParameters := map[string]string{"skuName": skuName, "maxShares": "3", "cachingMode": "None"}
 
```

Comment:
Hey,

I was looking at the test failure that you encountered, and the symptom seems like it was pretty similar to an issue I was facing earlier. I think that you need to set `VolumeBindingMode` to immediate, otherwise polling for the `multi-az` case doesn't work. The `PersistentVolume` for the volumes of that pod will be `nil`. Here's a snippet of the code that I had written to do that:



```go
if testutil.IsZRSSupported(location) {
	immediate := storagev1.VolumeBindingImmediate
	for i := range pod.Volumes {
		pod.Volumes[i].VolumeBindingMode = &immediate
	}
}
```

You have to add this in after the call to `NormalizeVolumes` because that changes the volume binding mode on the volumes of the pod. 

If this is indeed the issue, we might want to take a look at usage of `VerifySuccessfulAzVolumeAttachments` after this PR to see if the ```if volume.PersistentVolume != nil``` inside of it should actually be an error condition because it's tricky behavior. 

Joshua

Reply:
Hey Joshua. Sorry for the delay; I was addressing problems from other PR yesterday and wasn't able to get to this promptly.

And the problem indeed seems to be caused because of volume binding mode being set to `WaitForFirstCustomer` when it should be `VolumeBindingImmediate` as you suggested. For ZRS disks, `GetDynamicProvisioningStorageClass` function is responsible for setting the volume binding mode to `VolumeBindingImmediate`.
https://github.com/kubernetes-sigs/azuredisk-csi-driver/blob/04aba1e20ce3598d77e359198e4579335eb27024/test/e2e/driver/azuredisk_driver.go#L62-L64

The problem seems to be that the `SkuNameField` key in the storage class parameter is hard coded and is in camel case when it should not be (`SkuNameField` defined in azure constants package is all lower cased).
https://github.com/kubernetes-sigs/azuredisk-csi-driver/blob/04aba1e20ce3598d77e359198e4579335eb27024/test/e2e/dynamic_provisioning_test.go#L1652

Actually, the `SkuNameField` key value seems to be inconsistent throughout the entire test suite (some are camel cased and some aren't). So I have raised a PR (https://github.com/kubernetes-sigs/azuredisk-csi-driver/pull/1305) to address this as this make it difficult to detect bugs.

Reply:
Thanks for finding the root cause!

Reply:
I'm still seeing this issue after correcting the SKU field key. It seems that `NormalizeVolumes` sets the Volume's binding mode to `storagev1.VolumeBindingWaitForFirstConsumer` in the multi-az case. This results in the following check failing and the `PersistentVolume` field of `tpvc` never being set:

https://github.com/jmclong/azuredisk-csi-driver/blob/88123cd850a0082c59ba016d7eccdc04223a0f8e/test/resources/volumes.go#L101:L101

Reply:
That's probably because `SetUpDynamicPersistentVolumeClaim` relies on `VolumeDetail.BindingMode` instead of correctly set storage class's binding mode.
https://github.com/kubernetes-sigs/azuredisk-csi-driver/blob/04aba1e20ce3598d77e359198e4579335eb27024/test/resources/volumes.go#L101

Reply:
I just made the abovementioned changes and added it to the https://github.com/kubernetes-sigs/azuredisk-csi-driver/pull/1305 as a supplemental change.

And I have made a few changes in this PR to `VerifySuccessfulAzVolumeAttachments`, `SetupWithDynamicVolumes`, and `SetupWithDynamicMultipleVolumes` to address issues with `PodDetails.Volumes[n].PersistentVolume` being `nil`, which I will be pushing shortly.

Reply:
Nice! I was about to validate that solution. (using `storageClass.VolumeBindingMode` instead of `volume.VolumeBindingMode`).